jupyter_sphinx/execute.py

```diff
@@ -0,0 +1,373 @@
+from sphinx.transforms import SphinxTransform
+from sphinx.errors import ExtensionError
+from sphinx.addnodes import download_reference
+from sphinx.ext.mathbase import displaymath
+
+from docutils import nodes
+from IPython.lib.lexers import IPythonTracebackLexer, IPython3Lexer
+from docutils.parsers.rst.directives import flag, unchanged
+from docutils.parsers.rst import Directive
+
+import nbconvert
+from nbconvert.preprocessors.execute import executenb
+from nbconvert.preprocessors import ExtractOutputPreprocessor
+from nbconvert.writers import FilesWriter
+
+from jupyter_client.kernelspec import get_kernel_spec, NoSuchKernel
+
+import nbformat
+
+
+from ._version import __version__
+
+logger = logging.getLogger(__name__)
+
+
+def blank_nb(kernel_name):
+    try:
+        spec = get_kernel_spec(kernel_name)
+    except NoSuchKernel as e:
+        raise ExtensionError('Unable to find kernel', orig_exc=e)
+    return nbformat.v4.new_notebook(metadata={
+        'kernelspec': {
+            'display_name': spec.display_name,
+            'language': spec.language,
+            'name': kernel_name,
+        }
+    })
+
+
+def split_on(pred, it):
+    """Split an iterator wherever a predicate is True."""
+
+    counter = 0
+
+    def count(x):
+        nonlocal counter
+        if pred(x):
+            counter += 1
+        return counter
+
+    # Return iterable of lists to ensure that we don't lose our
+    # place in the iterator
+    return (list(x) for _, x in groupby(it, count))
+
+
+class Cell(nodes.container):
+    """Container for input/output from Jupyter kernel"""
+    pass
+
+
+def visit_container(self, node):
+    self.visit_container(node)
+
+
+def depart_container(self, node):
+    self.depart_container(node)
+
+
+class JupyterCell(Directive):
+
+    required_arguments = 0
+    final_argument_whitespace = True
+    has_content = True
+
+    option_spec = {
+        'hide-code': flag,
+        'hide-output': flag,
+        'code-below': flag,
+        'new-notebook': unchanged,
+        'kernel': unchanged,
+    }
+
+    def run(self):
+        self.assert_has_content()
+        if 'kernel' in self.options and 'new-notebook' not in self.options:
+            raise ExtensionError(
+                "In code execution cells, the 'kernel' option may only be "
+                "specified with the 'new-notebook' option."
+            )
+        # Cell only contains the input for now; we will execute the cell
+        # and insert the output when the whole document has been parsed.
+        return [Cell('',
+            nodes.literal_block(
+                text='\n'.join(self.content),
+                language='ipython'
+            ),
+            hide_code=('hide-code' in self.options),
+            hide_output=('hide-output' in self.options),
+            code_below=('code-below' in self.options),
+            kernel_name=self.options.get('kernel', '').strip(),
+            new_notebook=('new-notebook' in self.options),
```

Comment:
I am not sure about the `new_notebook` option, which seems to be used as a means to make code snippets run in separate execution contexts.

I would prefer something like `restart_kernel`. Using`notebook` in the name of the options seems to expose the fact that it is creating a notebook under the hood, which is an implementation detail IMO.

We could remove the feature altogether for a first version, since this may be confusing to users. Or we could havesomething like `kernel_id` which will specify the kernel id to be created or something like this.

Reply:
I agree about the naming, indeed. I also agree that under-the-hood creation of notebooks is not a good thing to advertise.

At the same time we do want to allow the users to easily extract and run the code, and also specify the filename. Does that make sense?

So the user may need to:
* Specify the filename
* Indicate that the new kernel is needed

Not entirely sure what's the best way to translate these two parameters into options, and how to call those.

Reply:
I am not sure what you mean about the file name?

Reply:
in the `jupyter-download:...` directives we allow to download the code as a notebook or a script. So we'd need to specify a filename for it to be human-readable.

Reply:
603ef72 introduces a `jupyter-kernel` directive that is used to specify the kernel to use:

```
.. jupyter-kernel:: python3
    :id: my-kernel-id
```

the `id` is used as the filename and the argument to the directive is the kernel name.